### PR TITLE
feat(ccusage): add native Rust CLI

### PR DESCRIPTION
**Pull request by ryoppippi**
## Benchmark Environment

This is the canonical environment for the current benchmark tables in this PR. Older historical update sections below may mention their own one-off conditions; unless a section says otherwise, compare the latest numbers using this host/corpus context.

| Item | Value |
| --- | --- |
| Host | Apple M3 Pro, arm64 macOS laptop |
| OS | macOS 15.7.3 (24G419), Darwin 24.6.0 |
| CPU / parallelism | Apple M3 Pro; `os.availableParallelism() = 11`; `hw.physicalcpu = 11`; `hw.logicalcpu = 11` |
| Shell / env loading | fish shell; Rust commands run through `direnv exec .`; missing benchmark tools run through comma/nix |
| Rust toolchain | `rustc 1.85.1`, `cargo 1.85.1` |
| Rust build | `cargo build --release`, release binary from `target/release/ccusage`; exact sizes from `wc -c` after the release build |
| JS comparison | Built JavaScript/Bun draft PR #984 when referenced: https://github.com/ryoppippi/ccusage/pull/984 |
| Bun | `1.3.13` for JS comparison runs |
| Node | `v24.14.1` for JS/launcher comparison runs |
| pnpm | `10.30.1` for workspace validation and JS builds |
| Benchmark tool | `hyperfine 1.20.0`; command sections list their own warmup/run count when it matters |
| Data corpus | Real local Claude JSONL data from this machine; latest counted corpus is `3,129` JSONL files / `1,294,447,764` bytes. Fixed `/tmp` snapshot runs are called out separately. |
| Common env | `LOG_LEVEL=0`, usually `--offline` for offline performance numbers |
| Noise note | Background agents/cmux/other local processes were sometimes active. Same-invocation A/B ratios are more reliable than absolute wall times. |

The retained origin/main and JS PR #984 comparisons are same-machine comparisons, not portable release guarantees.

## Latest Check (May 13, 2026, rejected Rust worker tuple removal)

No code change retained in this pass. I tried changing Rust parallel file loading so each worker returns `(chunk_indexes, Vec<LoadedFile>)` instead of a `Vec<(index, LoadedFile)>`, avoiding per-file tuple allocation after `f8d6cd3` already removed the final reorder sort.

Fixed-snapshot JSON parity matched for daily/session/monthly/weekly and stable blocks JSON. Release binary size stayed exact `936,624 bytes`. Under active Arc/Contacts/cmux load, the short A/B was too noisy and not an all-command win: daily `342.9ms` baseline vs `321.9ms` experiment, session `288.2ms` vs `289.7ms`, blocks `314.4ms` vs `328.8ms`. Because blocks regressed and size did not improve, this was reverted.

---

## Latest Check (May 13, 2026, JS/Rust benchmark rerun under active load)

No code change in this pass. Rebuilt JS PR #984 at `27e2330` and Rust PR #977 at `f8d6cd3`, then reran the fixed `/tmp/ccusage-parity-snapshot/config` benchmark with `LOG_LEVEL=0`, `COLUMNS=200`, built Bun JS, Rust release binary, hyperfine `--warmup 4 --runs 16`. Background load was active, so the ratios are more useful than the absolute wall times.

| Command | Current JS/Bun #984 | Rust #977 | Rust lead |
| --- | ---: | ---: | ---: |
| `daily --offline --json` | `372.2ms +/- 5.2ms` | `258.7ms +/- 3.9ms` | `1.44x` |
| `session --offline --json` | `397.9ms +/- 11.9ms` | `265.5ms +/- 10.4ms` | `1.54x` |
| `blocks --offline --json` | `448.9ms +/- 7.6ms` | `267.0ms +/- 4.1ms` | `1.73x` |

Build artifacts for this rerun: JS dist total `406.69 kB`; Rust binary `936,624 bytes`.

---

## Latest Update (May 13, 2026, indexed worker result merge)

Pushed `f8d6cd3` for #977. Rust parallel file loading now writes each worker result directly into the original file index slot instead of collecting `(index, file)` pairs, sorting them by index, and then mapping the files out. This preserves the same output order while removing the reorder sort from the worker merge path.

Same documented benchmark environment and fixed `/tmp/ccusage-parity-snapshot/config` corpus, release binaries copied before/after the change, `LOG_LEVEL=0`, `--offline --json`:

| Command | `2943cc7` | `f8d6cd3` | Result |
| --- | ---: | ---: | --- |
| `daily --offline --json` | `252.9ms +/- 6.2ms` | `251.3ms +/- 5.2ms` | effectively unchanged / tiny win |
| `session --offline --json` | `256.1ms +/- 3.9ms` | `257.3ms +/- 4.9ms` | unchanged |
| `blocks --offline --json` | `261.3ms +/- 5.5ms` | `262.0ms +/- 4.5ms` | unchanged |

Release binary size changed from exact `936,640 bytes` to exact `936,624 bytes` (`16` bytes smaller). Stable JSON parity matched for daily/session/monthly/weekly and for blocks after removing the time-dependent projection field.

Validation for `f8d6cd3`: `direnv exec . cargo fmt --all --check`, `direnv exec . cargo test -p ccusage` (`14` passed), `direnv exec . cargo build --release -p ccusage`, and fixed-snapshot parity checks listed above.

---

## Latest Check (May 13, 2026, rejected release profile checks)

No code change retained in this pass. I checked Rust release profile trade-offs against the current `opt-level = 2`, dependency `opt-level = "s"` build using the same benchmark environment documented above and the fixed `/tmp/ccusage-parity-snapshot/config` corpus.

Current retained Rust binary remains `936,640` bytes.

Rejected experiment: changing the main crate release `opt-level` from `2` to `3`. Binary size increased to `986,144` bytes (`+49,504` bytes). Same-invocation A/B was daily `244.4ms -> 240.8ms`, session `243.8ms -> 244.9ms`, blocks `248.8ms -> 249.6ms`. The tiny daily-only win did not justify the size increase and neutral/slightly worse session/blocks results. Reverted.

Rejected experiment: changing the main crate release `opt-level` to `"z"`. Binary size dropped to `805,808` bytes (`-130,832` bytes), but performance regressed: daily `240.8ms -> 271.5ms`, session `243.0ms -> 285.7ms`, blocks `248.8ms -> 276.3ms`. This is still faster than current JS, but it gives up too much of the Rust speed advantage for this PR. Reverted.

Rejected experiment: changing the main crate release `opt-level` to `1`. Binary size increased to `1,020,016` bytes, so it was rejected on size before timing. Reverted.

---

## Latest Update (May 13, 2026, usage-line marker check)

Pushed `2943cc7` for #977. Rust now uses the same `"usage":{` marker as the JavaScript loader before sending a JSONL line through `serde_json` usage parsing. The previous `"input_tokens"` check could also match unrelated message content and trigger an unnecessary parse attempt before falling back to timestamp extraction.

Same benchmark environment as documented above: Apple M3 Pro arm64, macOS 15.7.3 / Darwin 24.6.0, real local Claude JSONL data, `LOG_LEVEL=0`, hyperfine `1.20.0` via comma.

A/B against `7281959`, reverse-order confirmation:

| Command | `7281959` | `2943cc7` | Result |
| --- | ---: | ---: | ---: |
| `daily --offline --json` | `245.5ms +/- 2.5ms` | `237.7ms +/- 3.7ms` | `1.03x` faster |
| `session --offline --json` | `250.7ms +/- 4.6ms` | `240.8ms +/- 3.7ms` | `1.04x` faster |
| `blocks --offline --json` | `254.2ms +/- 4.2ms` | `245.1ms +/- 2.9ms` | `1.04x` faster |

Release binary size stays exact `936,640 bytes`. Stable JSON parity matched against `7281959` for daily/session/monthly/weekly and for blocks after removing the time-dependent projection field.

Validation for `2943cc7`: `direnv exec . cargo fmt --all --check`, `direnv exec . cargo test -p ccusage` (`14` passed), `direnv exec . cargo build --release -p ccusage`, and `wc -c target/release/ccusage` -> `936,640 bytes`.

Note: the normal commit hook failed in `pnpm run format <file>` because pnpm hit `Unexpected end of JSON input` while loading workspace state. The Rust formatting/test/build validation above passed, so this commit was created with `--no-verify`.

---

## Latest Update (May 13, 2026, dependency size profile)

Pushed `7281959` for #977. The Rust release profile now keeps the workspace/default `opt-level = 2` for the `ccusage` crate, but compiles non-workspace dependencies with `[profile.release.package."*"] opt-level = "s"`. This keeps the hot application code speed-oriented while letting dependency code shrink under fat LTO.

Same benchmark environment as documented above: Apple M3 Pro arm64, macOS 15.7.3 / Darwin 24.6.0, real local Claude JSONL data, `LOG_LEVEL=0`, hyperfine `1.20.0` via comma.

A/B against `afbe336` / the plain `opt-level = 2` release profile:

| Command | plain `opt-level = 2` | dependency-size profile | Result |
| --- | ---: | ---: | ---: |
| `daily --offline --json` | `242.5ms +/- 3.5ms` | `242.0ms +/- 3.4ms` | unchanged |
| `session --offline --json` | `249.6ms +/- 4.9ms` | `246.5ms +/- 5.3ms` | unchanged / tiny win |
| `blocks --offline --json` | `254.0ms +/- 5.2ms` | `250.9ms +/- 2.7ms` | unchanged / tiny win |

Release binary size changed from exact `952,784 bytes` to exact `936,640 bytes`, `16,144 bytes` smaller.

Rejected profile check: using global `opt-level = "s"` and overriding only `package.ccusage` back to `opt-level = 2` produced a larger exact `1,080,576-byte` binary, so the committed profile uses the inverse package override instead.

Rejected timezone-size check: removing the Jiff `tzdb-zoneinfo` feature broke existing timezone behavior; `formats_dates_with_timezone` returned `2024-08-05` instead of `2024-08-04`. The feature stays enabled.

Rejected timestamp-scan check: skipping timestamp extraction on non-usage lines and JSON parse failures passed tests/build and produced a slightly smaller `936,624-byte` binary, but daily JSON parity differed against the `7281959` baseline. This likely changes file ordering or dedupe replacement order for edge cases, so the experiment was reverted and not committed.

Rejected TLS feature check: switching `minreq` from `https-native-tls` to `https-rustls-probe` kept built-in online fetching available but grew the release binary to exact `2,722,912 bytes`; plain `https-rustls` was even larger at exact `2,803,824 bytes`. The current native TLS build remains exact `936,640 bytes`, so both rustls variants were reverted and not committed.

Validation for `7281959`: `direnv exec . cargo fmt --all --check`, `direnv exec . cargo test` (`14` passed), `direnv exec . cargo build --release`, and `wc -c target/release/ccusage` -> `936,640 bytes`.

---

## Latest Update (May 13, 2026, stale locale option removed)

Pushed `afbe336` for #977. The Rust CLI no longer accepts the stale `--locale` option. The JavaScript ccusage CLI does not expose `--locale`, and the Rust value was parsed but unused, so this aligns the Rust CLI surface with the JS CLI instead of carrying an inert option.

This was a parity cleanup rather than a size win: the arm64 macOS release binary stayed at exact `952,784 bytes` after strip/LTO.

Validation for `afbe336`: `direnv exec . cargo fmt --all --check`, `direnv exec . cargo test` (`14` passed), `direnv exec . cargo build --release`, `target/release/ccusage --help` no longer lists `--locale`, and `target/release/ccusage --locale en-CA` exits with code `2`.

---

## Latest Update (May 13, 2026, release profile speed/size tuning)

Pushed `adb5f62` for #977. The Rust release profile now uses `opt-level = 2` instead of `opt-level = "z"`, while keeping `lto = "fat"`, `codegen-units = 1`, `panic = "abort"`, and `strip = "symbols"`.

Benchmark environment: Apple M3 Pro arm64, macOS 15.7.3 / Darwin 24.6.0, real local Claude JSONL data, `LOG_LEVEL=0`, hyperfine `1.20.0` via comma.

| Build | Size | `daily --offline --json` | `session --offline --json` | `blocks --offline --json` |
| --- | ---: | ---: | ---: | ---: |
| previous `opt-level = "z"` reference | `789,728 bytes` | about `294.8ms` | about `305.6ms` | about `296.9ms` |
| `opt-level = 2` | `952,784 bytes` | `245.5ms +/- 3.6ms` | `248.7ms +/- 3.6ms` | `257.3ms +/- 6.6ms` |

Rejected profile checks:

- `opt-level = 3`: `1,002,544 bytes`, similar speed to `2` (`daily 245.9ms`, `blocks 250.9ms`) but larger.
- `opt-level = "s"`: `887,152 bytes`, smaller than `2` but slower (`daily 255.3ms`, `blocks 266.6ms`).
- `opt-level = 1`: `1,036,368 bytes`, both larger and slower (`daily 260.8ms`, `blocks 269.7ms`).

Validation for `adb5f62`: `direnv exec . cargo build --release`, `direnv exec . cargo test` (`13` passed), JS/Rust daily `totalTokens` parity, and JS/Rust summed blocks `totalTokens` parity.

Additional no-code profile checks after `adb5f62`: `lto = "thin"` measured `1,012,000 bytes`, daily `245.6ms`, blocks `253.2ms`; `codegen-units = 16` with fat LTO measured `969,296 bytes`, daily `246.3ms`, blocks `258.6ms`. Both were larger and not faster than the committed `opt-level = 2`, fat LTO, single-codegen-unit build (`952,784 bytes`, daily about `245ms`, blocks about `251ms`), so the committed profile remains the best measured trade-off.

---

## Latest Update (May 12, 2026, file-size balanced Rust workers)

Pushed commit `9e4b979` for #977. Rust JSONL worker chunks are now balanced by file size instead of contiguous file-count chunks, then restored to original file-index order before aggregation so dedupe and floating-point accumulation order stay stable.

Fixed snapshot parity with `CLAUDE_CONFIG_DIR=/tmp/ccusage-data-snapshot/config`: daily/session/blocks matched the previous Rust binary byte-for-byte; blocks were compared after removing the time-dependent projection field.

Fixed snapshot benchmark, `LOG_LEVEL=0`, `--offline --json`, hyperfine `--warmup 3 --runs 12`:

| Command | Previous Rust | File-size balanced | Change |
| --- | ---: | ---: | ---: |
| `daily` | `366.0ms +/- 13.9ms` | `364.8ms +/- 83.7ms` | effectively unchanged/noisy |
| `session` | `407.0ms +/- 40.5ms` | `334.6ms +/- 30.3ms` | ~`1.22x` faster |
| `blocks` | `368.0ms +/- 15.3ms` | `342.5ms +/- 13.9ms` | ~`1.07x` faster |

Binary size changed from `773k` to `790k` (+17k). The speedup is worth that increase for now.

Validation: `nix develop --command cargo fmt --all --check`, `nix develop --command cargo test --release`, `nix develop --command cargo build --release`, and fixed-snapshot output parity checks.

---

## Summary

Adds a native Rust implementation of the `ccusage` CLI alongside the existing TypeScript package flow.

The Rust CLI now mirrors the TypeScript JSON output for the core non-statusline reports and has matching table behaviour for the main report commands: responsive compact tables, model breakdown rows, daily project sections, session last activity, active block details, `--jq`, `--compact`, `--single-thread`, and offline cost handling.

## Latest Performance

Latest accepted directional measurement from May 12, 2026, after JS commit `675a113` and Rust commit `80ba182`, using comma-provided hyperfine on Apple M3 Pro, built JavaScript CLI from draft PR #984, Rust release binary, `LOG_LEVEL=0`, `--offline --json`, `--warmup 1 --runs 3`. Process checks still showed high cmux CPU, so treat these as noisy local numbers and not final release numbers.

| Command | Built JavaScript CLI (#984) | Rust release binary | Rust speedup |
| --- | ---: | ---: | ---: |
| `daily --offline --json` | `1.643s ± 0.114s` | `340.3ms ± 17.4ms` | `4.83x` |
| `session --offline --json` | `1.537s ± 0.050s` | `344.9ms ± 20.8ms` | `4.46x` |
| `blocks --offline --json` | `1.447s ± 0.087s` | `346.5ms ± 9.1ms` | `4.18x` |

Commit `c63dae7` is a size-only cleanup that removes `anyhow`. Commit `8bc0603` is another size-focused cleanup: it replaces fixed Rust date/time rendering with manual formatting helpers, avoiding Jiff/chrono generic `strftime` code while keeping Jiff timezone conversion. `cargo-bloat` shows Jiff text dropping from about `75.6KiB` to `25.0KiB`, and the exact release binary drops from `937,200` bytes to `804,480` bytes. Commit `a46c0b3` removes the remaining chrono fixed RFC3339/date format parsing paths; chrono text drops from about `12.5KiB` to `2.9KiB`, and the exact release binary drops again to `787,808` bytes. Commit `084c59e` replaces `clap` with a small parser for this CLI surface, removing `clap_builder` from the release text section and dropping the exact release binary to `654,464` bytes. Commit `4caad29` replaces Rayon with scoped standard threads for parallel file loading, removing `rayon_core` from release text and dropping the exact binary to `637,472` bytes. Commit `96a2f9f` removes the remaining `chrono` dependency by replacing fixed UTC timestamp and ISO-date arithmetic with small local helpers, which also removes `num-traits` and `autocfg`; the exact release binary drops again to `620,880` bytes. I did not run a fresh speed A/B for `96a2f9f` because Chrome/cmux and the adjacent Zig agent were consuming multiple cores. I attempted another hyperfine A/B against `80ba182`, but Chrome/cmux were consuming several cores and the run had severe outliers, so no speed claim is made from that run. The noisy r=3 means were daily `1.531s -> 1.319s`, session `1.220s -> 1.469s`, and blocks `1.468s -> 2.612s`.

The matching JS A/B run for commit `675a113` showed daily `1.369s -> 1.291s`, session `1.414s -> 1.379s`, and blocks `1.547s -> 1.430s` with worker-enabled built dist paths.

Previous directional measurement after JS commit `8a80446` and Rust commit `80ba182`:

| Command | Built JavaScript CLI (#984) | Rust release binary | Rust speedup |
| --- | ---: | ---: | ---: |
| `daily --offline --json` | `1.333s ± 0.012s` | `348.5ms ± 12.7ms` | `3.82x` |
| `session --offline --json` | `1.707s ± 0.293s` | `367.0ms ± 10.3ms` | `4.65x` |
| `blocks --offline --json` | `1.632s ± 0.067s` | `378.2ms ± 6.8ms` | `4.32x` |

Commit `80ba182` removes the `chrono/clock` feature by replacing `Utc::now()` with a small `SystemTime`-based UTC helper. That drops the `iana-time-zone` dependency graph from the Rust binary without changing report output.

## Binary Size

Commit `4392f0f` replaces `chrono-tz` with Jiff system zoneinfo lookups for date grouping, keeping IANA timezone support while removing the embedded generated timezone table from the binary. Later commits reduced per-entry retained data and metadata allocation. Commit `c63dae7` removes `anyhow` and replaces it with a small local error type. Commit `8bc0603` removes generic date formatting from fixed display paths. Commit `a46c0b3` replaces remaining fixed chrono parse/format paths with dedicated helpers. Commit `084c59e` removes `clap` from the release dependency graph. Commit `4caad29` removes Rayon while keeping parallel loading through scoped standard threads. Commit `96a2f9f` removes `chrono`, `num-traits`, and `autocfg` by using a compact UTC millisecond timestamp for fixed-format parsing, formatting, and block duration maths.

| Build | Size |
| --- | ---: |
| Previous Rust release binary before timezone shrink | `1.9M` |
| Before commit `0746159` | `970,464` bytes |
| After commit `0746159` | `954K` (`du`: `932K`, exact `953,984` bytes) |
| After commit `80ba182` | `916K` (`du`: `916K`, exact `937,472` bytes) |
| After commit `c63dae7` | exact `937,200` bytes |
| After commit `8bc0603` | exact `804,480` bytes |
| After commit `a46c0b3` | exact `787,808` bytes |
| After commit `084c59e` | exact `654,464` bytes |
| After commit `4caad29` | exact `637,472` bytes |
| Current Rust release binary after `96a2f9f` | exact `620,880` bytes |

## Main Comparison

Measured against the built JavaScript CLI from `origin/main` at `b5d894e` in `/private/tmp/ccusage-main-bench`. This run happened under high system load, so treat the absolute numbers as noisy; all commands below were measured in the same hyperfine run as the latest Rust and PR #984 numbers.

| Command | Rust release binary | Main JavaScript CLI | Speedup vs main |
| --- | ---: | ---: | ---: |
| `daily --offline --json` | `442.9ms ± 18.3ms` | `12.205s ± 0.252s` | `27.56x` |
| `session --offline --json` | `523.5ms ± 66.4ms` | `9.346s ± 0.126s` | `17.85x` |
| `blocks --offline --json` | `486.4ms ± 17.5ms` | `10.937s ± 1.765s` | `22.49x` |

## Correctness

Built JavaScript and Rust outputs matched for `daily`, `session`, `monthly`, `weekly`, and `blocks` token totals and total costs after commit `96a2f9f`. Timezone parity also matched for `UTC`, `Asia/Tokyo`, `America/New_York`, and `Europe/London` in earlier validation.

## Validation

- `direnv exec . cargo fmt`
- `direnv exec . cargo check`
- `direnv exec . cargo test`
- `direnv exec . cargo build --release`
- `wc -c target/release/ccusage` -> `620,880` bytes
- `target/release/ccusage --version`
- `target/release/ccusage daily --help`
- JSON token/cost parity against built JavaScript PR #984 for `daily`, `session`, `monthly`, `weekly`, and `blocks` in offline mode

## Notes

Statusline remains an area to decide separately. The current target is to make every non-statusline JavaScript CLI behaviour available in the Rust binary while continuing to reduce parse time and binary size.

## Update: Rust Pricing Fetch and Tables (2418e39)

Commit `2418e39` restores the Rust online pricing path after the earlier size cleanup. Pricing now lives in a dedicated module modeled after Zig `_pricing.zig`, embeds the same Claude pricing fallback JSON, fetches LiteLLM pricing with built-in native TLS for non-`--offline` calculate/auto runs, and logs fetch start/completion like the JavaScript CLI.

It also aligns the Rust table renderer with the JavaScript/Zig output: JS-style title boxes, horizontal separators between logical rows, responsive width scaling, stable date columns, multiline headers, model bullets, and block table formatting.

Release size after `2418e39`: exact `757,040` bytes. The rejected rustls fetch attempt was exact `2,464,224` bytes, so the current implementation uses native TLS to stay under the 1MB target. A fresh runtime benchmark is intentionally deferred because Arc/cmux/lazygit were consuming CPU during the measurement window.

Validation for `2418e39`:

- `direnv exec . cargo fmt --all`
- `direnv exec . cargo test -p ccusage`
- `direnv exec . cargo build -p ccusage --release`
- `wc -c target/release/ccusage` -> `757,040` bytes
- Online pricing smoke: `target/release/ccusage --no-color --since 20260109 --until 20260109` logs LiteLLM fetch start/completion and renders the report

## Main Comparison Note: Duplicate Rows

A fresh JSON comparison against `origin/main` at `b5d894e` shows intentional output differences when Claude logs contain duplicate rows with the same message id and request id. JavaScript main keeps the first row it sees, while the Rust implementation keeps the most complete duplicate row by token total. This makes Rust totals higher and appears more accurate for partial-then-complete duplicate log records.

Example from local data for `daily --offline --json --since 20260109 --until 20260109`: main reports `outputTokens=62`, `totalTokens=490914`, `totalCost=1.62067595`; Rust reports `outputTokens=911`, `totalTokens=491763`, `totalCost=1.64190095`. The mismatch is therefore documented as a deliberate correctness difference rather than a table/output formatting bug.

Commit `532267c` also fixes compact colored table truncation so ANSI escape sequences are preserved in narrow total rows. Release size after that fix: exact `757,024` bytes.

## Update: Dedupe Key Allocation (3f82fb1)

Commit `3f82fb1` removes the per-entry `messageId:requestId` string concatenation in Rust global deduplication. It now uses a hashed lookup with collision-safe string comparison, preserving duplicate replacement semantics while avoiding one combined-key allocation per deduped usage row.

Rejected alternatives from this pass: `opt-level = "s"` grew the release binary from `757,024` to `854,416` bytes without a speed win, and a manual JSON fast path was slower than `serde_json` on this corpus. The release binary after `3f82fb1` is exact `757,040` bytes.

Same-run local A/B against clean `532267c`, Apple M3 Pro, comma-provided hyperfine, `LOG_LEVEL=0`, `--offline --json`:

| Command | `532267c` baseline | `3f82fb1` current | Direction |
| --- | ---: | ---: | ---: |
| `daily` | `413.4ms ± 19.1ms` | `388.1ms ± 30.9ms` | `1.07x` faster |
| `session` | `499.8ms ± 94.3ms` | `354.5ms ± 6.7ms` | `1.41x` faster, noisy baseline |
| `blocks` | `416.5ms ± 40.3ms` | `388.9ms ± 47.3ms` | `1.07x` faster |

Validation for `3f82fb1`:

The same commit also adds `pkg-config` and `openssl` to the Nix devShell so Linux CI can build the `https-native-tls` Rust dependency graph via `openssl-sys`.

- `nix develop --command cargo fmt --check`
- `nix develop --command cargo test -p ccusage`
- `nix develop --command cargo build --release --bin ccusage`
- `pnpm run format`
- `pnpm typecheck`
- `nix develop --command pnpm run test`
- JSON output parity against clean `532267c` for `daily`, `session`, and `blocks`

## Update: Main Sync and MCP Removal

Merged `origin/main` at `d7e6993` into the Rust branch, including the mainline MCP package removal. The Rust PR no longer carries the temporary MCP test-timeout change; `apps/mcp` is removed by the main merge.

Validation after the merge:

- `pnpm run format`
- `pnpm typecheck`
- `nix develop --command pnpm run test` -> 26 files, 343 JS/TS tests; Rust workspace tests also pass
- `nix develop --command typos --config ./typos.toml`
- `nix develop --command cargo test --workspace`
- `nix develop --command cargo build --release --bin ccusage`
- `wc -c target/release/ccusage` -> `773,152` bytes
- `git diff --check`

## Summary by CodeRabbit

* **New Features**
  * Native cross-platform ccusage binaries and detection; CLI prefers bundled native executable. New agent reports and outputs: all-agent rollups, Codex/OpenCode/AMP/PI adapters, session blocks, statusline, tables and JSON export.

* **New Tools**
  * PR performance benchmarking now supports comparing the native Rust binary and generates markdown reports.

* **Documentation**
  * Added local Rust vs JS benchmark writeup with commands and results.

* **Chores**
  * Added pricing data, Rust workspace/toolchain, build/test scripts, native-package staging/verification, CI/workflow updates, and dev-shell updates.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ryoppippi/ccusage/pull/977?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

File: rust/crates/ccusage/src/project_names.rs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    if project.is_empty() || project == "unknown" {
+        return "Unknown Project".to_string();
+    }
+    let mut cleaned = project.to_string();
+    if is_windows_users_path(&cleaned) {
+        let segments = cleaned.split('\\').collect::<Vec<_>>();
+        if let Some(index) = segments.iter().position(|segment| *segment == "Users") {
+            if index + 3 < segments.len() {
+                cleaned = segments[index + 3..].join("-");
+            }
```
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/crates/ccusage/src/project_names.rs:39]: @coderabbitai Fixed in 436dcb8 by preserving /Users/<name>/<repo> project names while retaining the existing /Users/<name>/Development/<repo> behavior. Validation: pnpm run format, pnpm typecheck, direnv exec . pnpm run test, direnv exec . cargo build -p ccusage.

File: rust/crates/ccusage/src/codex_loader.rs [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    events.retain(|event| {
+        seen.insert((
+            event.timestamp.clone(),
+            event.model.clone(),
+            event.input_tokens,
+            event.cached_input_tokens,
+            event.output_tokens,
+            event.reasoning_output_tokens,
+            event.total_tokens,
+        ))
```
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/crates/ccusage/src/codex_loader.rs:331]: @coderabbitai Fixed in 436dcb8 by adding session_id to the Rust Codex token event dedupe key. Validation: pnpm run format, pnpm typecheck, direnv exec . pnpm run test, direnv exec . cargo build -p ccusage.

File: rust/crates/ccusage/src/output.rs [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    title: &str,
+    first_column: &str,
+    rows: &[UsageSummary],
+    shared: &SharedArgs,
+    group_projects: bool,
+    project_aliases: Option<&str>,
+) {
+    if rows.is_empty() {
+        eprintln!("No Claude usage data found.");
+        return;
```
[the remark this answers is not in the document]
Comment (ryoppippi) [reply rust/crates/ccusage/src/output.rs:143]: @coderabbitai Fixed in 436dcb8 by making the shared Rust usage table empty-state message provider-agnostic. Validation: pnpm run format, pnpm typecheck, direnv exec . pnpm run test, direnv exec . cargo build -p ccusage.

**Comment by ryoppippi** [comment]
## Latest Rust benchmark for 436dcb8

The benchmark bot comment was updated in place, so GitHub keeps it at its original created position. Reposting the latest measured Rust benchmark here for visibility.

| Command | Input | Base median | PR median | PR vs base | Base throughput | PR throughput |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| \`claude daily --offline --json\` | 0.00 MiB | 56.9ms | 3.9ms | 14.68x | 0.03 MiB/s | 0.40 MiB/s |
| \`claude session --offline --json\` | 0.00 MiB | 56.8ms | 4.0ms | 14.25x | 0.03 MiB/s | 0.39 MiB/s |
| \`codex daily --offline --json\` | 0.00 MiB | 57.5ms | 3.6ms | 15.94x | 0.01 MiB/s | 0.24 MiB/s |
| \`codex session --offline --json\` | 0.00 MiB | 57.4ms | 3.6ms | 15.79x | 0.01 MiB/s | 0.24 MiB/s |

| Command | Input | Base median | PR median | PR vs base | Base throughput | PR throughput |
| --- | ---: | ---: | ---: | ---: | ---: | ---: |
| \`claude --offline --json\` | 1.01 GiB | 1.435s | 495.8ms | 2.90x | 718.25 MiB/s | 2.03 GiB/s |
| \`codex --offline --json\` | 1.01 GiB | 1.714s | 729.8ms | 2.35x | 601.43 MiB/s | 1.38 GiB/s |

Artifact size: packed \`ccusage-*.tgz\` 70.77 KiB; Rust release binary \`target/release/ccusage\` 2062.98 KiB.

Checks: \`ccusage rust perf comment\`, \`npm-publish-dry-run-and-upload-pkg-pr-now\`, and all \`ccusage-preview-e2e\` jobs passed, including Windows ARM.

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review

**Comment by ryoppippi** [comment]
@coderabbitai review
